New places api support

rojiphil opened the pull request:
@rafecolton @mkhutornyi  

### Details

This PR implements the [scope](https://github.com/Expensify/App/issues/45432#issuecomment-2263794353) limited to the places migration as used in NewDot.

Dependencies:
1. Changes in `react-native-google-places-autocomplete` library which is implemented in [this PR](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943)
2. API Key updation for `Places API (New)` in Google Cloud 
3. BE changes to support API changes for `Autocomplete (New)` and `Place Details(New)`

The implementation details for the relevant steps as mentioned [here](https://github.com/Expensify/App/issues/45432#issuecomment-2263739363) are as follows:

Let us introduce a `isNewPlacesAPI` property in `GooglePlacesAutocomplete` component [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1006C3-L1006C17) to introduce support for `Places (New)` api. By default, we can set this to `false` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1006C3-L1006C17) so that the current implementation of `Places` api is used by default. This way, we do not break existing systems using the old `Places` API. Further, to make use of the `Places API (New)`, the FE can send `isNewPlacesAPI` [here](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/components/AddressSearch/index.tsx#L476). 

1. Use HTTP POST instead of HTTP GET for Autocomplete (New) API uses as mentioned [here](https://developers.google.com/maps/documentation/places/web-service/place-autocomplete#place_autocomplete_requests).

For `Autocomplete (New)` request, we can send a HTTP POST request [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L609) and also set the new url for POST i.e. `${url}/v1/places:autocomplete`. Further, we can send the request `body` with the supported parameters [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L624C1-L630C12).
In addition, we also need to update the API for `Place Details (New)` with url `${url}/v1/places/${rowData.place_id}` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L334)

2. Handle the format changes in response structure for Autocomplete (New) API as mentioned [here](https://developers.google.com/maps/documentation/places/web-service/place-autocomplete#:~:text=5000%20meter%20radius%3A-,%7B,-%22suggestions%22%3A)

Format changes include changing the request parameters to `languageCode`, `includedRegionCodes` and `locationBias` [here](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/components/AddressSearch/index.tsx#L99C1-L101C49). For the response though, we need to filter the results and handle the format changes. This is done [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L440C1-L461C5) and [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L582C1-L589C12)

3. Use the required parameter fieldmask for Place Details (New) API as mentioned [here](https://developers.google.com/maps/documentation/places/web-service/place-details#fieldmask). This way we can control the billing by including what we need in field mask

Let us introduce a `fields` property in `GooglePlacesAutocomplete` component [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1007) to specify the desired response fields. By default, we can set this to `*` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L1053) so that all the fields are sent back. In our case, we set `fields` to make use of [Location Only (Place Details) SKU](https://developers.google.com/maps/documentation/places/web-service/usage-and-billing#location-placedetails)  thereby reducing our API cost. The specific value of `fields` is set [here](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/CONST.ts#L1193).

4. Use session tokens to group autocomplete and place details which will help in reduced billing. This is referenced [here](https://developers.google.com/maps/documentation/places/web-service/place-session-tokens).

To use sessions, we can create a `sessionToken` state [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L163) and use `uuid` library to generate v4 uuids for the session. This session token will be used in every `Autocomplete (New)` API [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L627). And the session token will be reset to a new uuid [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L340) after using it for the last time in the `Place Details (New)` API [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L336) 

5. Handle the format changes related to Place object for in response structure of Place Details (New) API as referenced [here](https://developers.google.com/maps/documentation/places/web-service/migrate-response)

Since the `response` has changed `Place Details (New)` API, there will not be any `status`. Hence, we can depend on `id` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/e3676609bae828a98ce4a1de44bf6312a968a4e5/GooglePlacesAutocomplete.js#L288) and pass the `details` to the `onPress` handler. The FE [saveLocationDetails](https://github.com/rojiphil/App/blob/6425853eadcd8a6c19a183217d97c102219b7416/src/components/AddressSearch/index.tsx#L106) is modified to support the format changes in the `Place Details (New)` API response.

### Fixed Issues
$ https://github.com/Expensify/App/issues/46771
PROPOSAL: 

### Test

**Steps:** 
1. Click on the Global FAB —> Submit Expense
2. Click on the `Distance` tap and click `Start` to setup a waypoint
3. Enter the search text in `Address` input field
4. Verify that the autocomplete results are displayed with the waypoint name and waypoint address
5. Click on one of the results and verify that the selected waypoint name is added as a `start` waypoint
6. Click `Stop` to setup a waypoint
7. Select an already predefined waypoint and verify that the selected waypoint is added as a `stop` waypoint

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as the steps in Tests Section.

### QA Steps

Same as the steps in Tests Section. 

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.


### Screenshots/Videos
<details>
<summary>MacOS: Safari</summary>

TODO

</details>

<details>
<summary>iOS: mWeb Safari</summary>

TODO

</details>

<details>
<summary>MacOS: Desktop</summary>

TODO

</details>

<details>
<summary>iOS: Native</summary>

TODO

</details>

<details>
<summary>Android: Native</summary>

TODO

</details>

<details>
<summary>Android: mWeb Chrome</summary>

TODO

</details>

rojiphil committed:
new places api support

rojiphil committed:
pass fields as property

rojiphil commented:
@rafecolton @mkhutornyi Please review the implementation details in OP. Also, please note the following:
1. Currently, as BE is not yet ready, error code `666` would be received. So, until the BE is ready to process the modified URLs, I have temporarily added `Autocomplete (New)` response [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/c7c5db69df63f2801e2a5871618dd48707cba318/GooglePlacesAutocomplete.js#L737C1-L811C12) and `Place Details (New)` API response [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/c7c5db69df63f2801e2a5871618dd48707cba318/GooglePlacesAutocomplete.js#L286C1-L453C12) so that we can test. These changes would be removed once the BE changes are ready.  
2. Although in development, we can still test this out by copying `GooglePlacesAutocomplete.d.ts` and `GooglePlacesAutocomplete.js` from the PR branch [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943) to the node_modules/react-native-google-places-autocomplete.

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rafecolton reviewed (commented):
Can you please merge main and update to use your branch of `react-native-google-new-places-autocomplete`? Will make it a little easier for me for testing

**Edit:** You can disregard this, I completed testing.

rafecolton commented on src/components/AddressSearch/index.tsx:
```diff
@@ -97,26 +96,25 @@ function AddressSearch(
-                    name: details.name,
+                    lat: details.location?.latitude ?? details.geometry.location?.lat ?? 0,
+                    lng: details.location?.longitude ?? details.geometry.location?.lng ?? 0,
+                    name: autocompleteData?.structured_formatting?.main_text ?? details?.name ?? '',
```
`details.name` isn't the same as it was in the old API, the format here is `places/<id>` so not really good to use here. The next closest thing is `details?. shortFormattedAddress`

rojiphil replied:
In this context, the `details.name` is not used from old API. Instead, it indicates a predefined option i.e. saved places in our app. So, I think we need to keep 
 `details.name`

rojiphil replied:
> I don't understand what you mean, can you show me?

Please find the test video that demonstrates the use of `predefined` places. This is the reason why I think we need to keep `details.name`.


https://github.com/user-attachments/assets/0ad30e8b-90e9-42b5-809d-bb4cd712f533

rafecolton commented on src/components/AddressSearch/index.tsx:
```diff
@@ -174,7 +172,7 @@ function AddressSearch(
         const values = {
             street: `${streetNumber} ${streetName}`.trim(),
-            name: details.name ?? '',
+            name: autocompleteData?.structured_formatting?.main_text ?? details?.name ?? '',
```
Same comment here as above

rojiphil replied:
Same as above

rafecolton reviewed (changes requested):
Everywhere we use `long_name` or `short_name` in both code and comments in `src/components/AddressSearch/index.tsx` and `src/libs/GooglePlacesUtils.ts`, it needs to be changed to `longName` and `shortName` respectively. With that small change, this works great!

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil commented:
Sorry for the delay here. I will look into this today and revert.

rojiphil commented:
> @rojiphil @mkhutornyi Are you able to test on Expensify Staging? If not, are you able to get your own API key so you can test e2e locally?

@rafecolton I am not able to test as I still get `666` response on accessing the new places API through the Staging server. Here is a video demonstrating this. What am I missing here? 

https://github.com/user-attachments/assets/4f7154ac-b5d3-4442-9c4c-e8b01ed8be34

rafecolton commented:
Ah, looks like you tested a couple hours before the code was on staging. Sorry for the confusion, I was asking in general. It's on staging now, so please try again and LMK.

rojiphil committed:
update rectangle object for locationbias

rojiphil commented:
> Ah, looks like you tested a couple hours before the code was on staging. Sorry for the confusion, I was asking in general. It's on staging now, so please try again and LMK.

Yeah @rafecolton. The BE changes are reflected in staging as I do not get `666` now. But I ran into the following problem. 

While I can fetch the `autocomplete` place predictions, I cannot fetch the `place details`. Strangely, it returns 200 but has no response contents. I am running out of ideas on why it does not return the response. I however can see that the API returns a valid response in the `Google Maps Platform`. I have also attached a test video to demonstrate all this.

Can you please help confirm in BE if the `place details` proxy request from FE returns a valid response from the actual `Place Details` API request made with the google platform?

https://github.com/user-attachments/assets/10730b82-443d-42e7-977e-4e79ffef11c9

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rafecolton commented:
That makes sense. See my prior review.

rojiphil commented:
> That makes sense. See my prior review.

@rafecolton  Sorry if I am missing something in your review here but can you please point out which one? 
As I understand, passing `languageCode` and `regionCode` as mentioned [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943#discussion_r1763981295) is optional. Also, I assume that `key` would get added in BE as we would not want to expose the key in FE. Further, the comment [here](https://github.com/Expensify/App/pull/47085#pullrequestreview-2311218844) deals with the `response` but we are not there yet. The `place details` API request itself is sending back empty response which baffles me.

rafecolton commented:
If you're seeing the API return data and the UI is not working, that means we're parsing it wrong in the front end. You can print it with `console.log` and compare it to the code to see what's wrong. 

That's what I did when I posted https://github.com/Expensify/App/pull/47085#pullrequestreview-2311218844, which explains what to do to fix the error you're seeing.

rojiphil reviewed (commented):
> it needs to be changed to `longName` and `shortName` respectively

Instead of `longName` and `shortName` the response now consists of `longText` and `shortText`. Looks like the response format has changed as can be seen [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/0d2614bb6827b804e19ce7485b3e8e37532e2a9a/GooglePlacesAutocomplete.js#L290C1-L376C16). I got this response when requesting via API explorer on Google Maps platform. With this data, the FE works well. 

To test this, please use `Statue of Liberty` as search text and select `Statue of Liberty National Monument` to request the place details api. For testing purposes, I am constructing a response when BE sends back empty response as referenced  [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/0d2614bb6827b804e19ce7485b3e8e37532e2a9a/GooglePlacesAutocomplete.js#L285C2-L290C29) and it works well. Here is a test video for this.


https://github.com/user-attachments/assets/d955d2f3-cfc6-4137-a8c8-336c0cb145ae

rojiphil commented:
> so once you make the changes I requested here, I'll test again against the BE proxy locally to see what's up.

@rafecolton I have addressed the review comments. Hopefully we will unravel the mystery of empty response when we test the BE proxy locally. 

I have added the `key` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/0d2614bb6827b804e19ce7485b3e8e37532e2a9a/GooglePlacesAutocomplete.js#L429) to address the review comment [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943#discussion_r1763979557). But looks like we need not worry about this as BE will add the `key` on receiving the proxy request from FE. 

I have added `languageCode` and `regionCode` [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/0d2614bb6827b804e19ce7485b3e8e37532e2a9a/GooglePlacesAutocomplete.js#L432-L433) to address the review comment [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943#discussion_r1763981295)

rojiphil commented:
@rafecolton While we address the issue of `placeDetails`, a different issue related to `AutoComplete` API request is also noticed. Let me also mention that here so that when you look into this issue, both can be covered.

This issue related to `AutoComplete` will occur when we have a valid `locationBias` to send. This will happen only after we select any location as a waypoint. Once added as a waypoint in our app, further `autocomplete` requests will contain `locationBias`. However, this will fail due to `Invalid JSON payload` error as demonstrated below. It looks like stringifying the JSON is causing an issue here.

As our current focus is to address the `PlaceDetails` issue, I have suppressed the error due `locationBias` by excluding from the parameter [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/0d2614bb6827b804e19ce7485b3e8e37532e2a9a/GooglePlacesAutocomplete.js#L721). To trigger this issue though, we can include the `locationBias` again.


https://github.com/user-attachments/assets/5dde5501-e815-4306-8cd3-ccb777bc3916

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil commented:
> Hopefully we will unravel the mystery of empty response when we test the BE proxy locally.

@rafecolton I just tested `placedetails` API now and, woohoo, it works great. So, we can ignore the comment [here](https://github.com/Expensify/App/pull/47085#issuecomment-2368038501). I have also removed the test code in the library as it is no longer required. The following test video demonstrates the working of placedetails API via proxy URL.

https://github.com/user-attachments/assets/aafa6942-c8f3-42ba-af4d-6016869b2a9e

rojiphil commented:
@rafecolton So the only pending issue is the comment [here](https://github.com/Expensify/App/pull/47085#issuecomment-2375989711) concerning `autocomplete(new)` api request via proxy. Wondering if a change in `content-type` is needed for the `autocomplete(new)` api.

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rafecolton reviewed (commented):
Can you please update this to use your branch https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943 until it's merged? It's really confusing to run this code locally and discover it's using the old API

rafecolton commented:
> Instead of longName and shortName the response now consists of longText and shortText. Looks like the response format has changed as can be seen [here](https://github.com/rojiphil/react-native-google-new-places-autocomplete/blob/0d2614bb6827b804e19ce7485b3e8e37532e2a9a/GooglePlacesAutocomplete.js#L290C1-L376C16). I got this response when requesting via API explorer on Google Maps platform. With this data, the FE works well.

Yes you are correct, it's `longText` not `longName` and `shortText` not `shortName`. You still haven't made this change to your code, so it's still not working. Any particular reason you still have not made that change yet? Here is the specific change required to make this work:

```patch
diff --git a/src/components/AddressSearch/index.tsx b/src/components/AddressSearch/index.tsx
index 712c364b646..f18c6c63272 100644
--- a/src/components/AddressSearch/index.tsx
+++ b/src/components/AddressSearch/index.tsx
@@ -135,27 +135,27 @@ function AddressSearch(
             country: countryPrimary,
         } = GooglePlacesUtils.getAddressComponents(addressComponents, {
             // eslint-disable-next-line @typescript-eslint/naming-convention
-            street_number: 'long_name',
-            route: 'long_name',
-            subpremise: 'long_name',
-            locality: 'long_name',
-            sublocality: 'long_name',
+            street_number: 'longText',
+            route: 'longText',
+            subpremise: 'longText',
+            locality: 'longText',
+            sublocality: 'longText',
             // eslint-disable-next-line @typescript-eslint/naming-convention
-            postal_town: 'long_name',
+            postal_town: 'longText',
             // eslint-disable-next-line @typescript-eslint/naming-convention
-            postal_code: 'long_name',
+            postal_code: 'longText',
             // eslint-disable-next-line @typescript-eslint/naming-convention
-            administrative_area_level_1: 'short_name',
+            administrative_area_level_1: 'shortText',
             // eslint-disable-next-line @typescript-eslint/naming-convention
-            administrative_area_level_2: 'long_name',
-            country: 'short_name',
+            administrative_area_level_2: 'longText',
+            country: 'shortText',
         });
 
-        // The state's iso code (short_name) is needed for the StatePicker component but we also
-        // need the state's full name (long_name) when we render the state in a TextInput.
+        // The state's iso code (shortText) is needed for the StatePicker component but we also
+        // need the state's full name (longText) when we render the state in a TextInput.
         const {administrative_area_level_1: longStateName} = GooglePlacesUtils.getAddressComponents(addressComponents, {
             // eslint-disable-next-line @typescript-eslint/naming-convention
-            administrative_area_level_1: 'long_name',
+            administrative_area_level_1: 'longText',
         });
 
         // Make sure that the order of keys remains such that the country is always set above the state.
diff --git a/src/libs/GooglePlacesUtils.ts b/src/libs/GooglePlacesUtils.ts
index 312a0dc61c1..147d6933435 100644
--- a/src/libs/GooglePlacesUtils.ts
+++ b/src/libs/GooglePlacesUtils.ts
@@ -1,8 +1,8 @@
 type AddressComponent = {
     // eslint-disable-next-line @typescript-eslint/naming-convention
-    long_name: string;
+    longText: string;
     // eslint-disable-next-line @typescript-eslint/naming-convention
-    short_name: string;
+    shortText: string;
     types: string[];
 };
 type FieldsToExtract = Record<string, Exclude<keyof AddressComponent, 'types'>>;
@@ -11,8 +11,8 @@ type FieldsToExtract = Record<string, Exclude<keyof AddressComponent, 'types'>>;
  * Finds an address component by type, and returns the value associated to key. Each address component object
  * inside the addressComponents array has the following structure:
  * [{
- *   long_name: "New York",
- *   short_name: "New York",
+ *   longText: "New York",
+ *   shortText: "New York",
  *   types: [ "locality", "political" ]
  * }]
  */
```

rafecolton commented:
Looking at the waypoint issue now

rafecolton commented:
Was able to reproduce the location bias issue. You are right, it looks like the problem is Google isn't handling URI-encoded data that's multiple layers deep, even though it's encoded correctly. Experimenting a little to see if it's better to send as JSON from the front-end or decode in PHP. Probably the former so this works for people not using a proxy

rafecolton reviewed (changes requested):
PR for back-end changes is up for review. I tested and confirmed it works with the diff from my review [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/943#pullrequestreview-2339036740) and the diff posted in [this comment](https://github.com/Expensify/App/pull/47085#issuecomment-2384186958). It fixes the locationBias issue you described. Please make both of those changes, and I will let you know when it's ready to test on staging.

Please also reply to my comment [here](https://github.com/Expensify/App/pull/47085#discussion_r1781623548) 🙏

This is almost ready! Thank you for your persistence and diligence in checking edge cases.

rojiphil commented:
>  It fixes the locationBias issue you described. Please make both of those changes, and I will let you know when it's ready to test on staging.

Yeah! Looks like this is pretty close to getting done. Thanks. I will pick this up tomorrow my day for making the necessary changes.

rafecolton commented:
The BE fix has been deployed to staging so you should be good to test. 

Do you think we can get the PRs ready for review today?

rojiphil committed:
using longText and shortText + prettier fix

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil commented:
> You still haven't made this change to your code, so it's still not working. Any particular reason you still have not made that change yet? Here is the specific change required to make this work:

Yeah. That got missed out. The required changes are made and thanks for pointing this out.

rojiphil commented:
> Please also reply to my comment [here](https://github.com/Expensify/App/pull/47085#discussion_r1781623548) 🙏

I have replied to the comment [here](https://github.com/Expensify/App/pull/47085#discussion_r1786135061)

rojiphil commented:
> The BE fix has been deployed to staging so you should be good to test.
> 
> Do you think we can get the PRs ready for review today?

@rafecolton The BE fix works great with staging along with the fixes in FE. 
I have bumped up the version in library and have placed the library in `ready for review` state.  Once released, I will test this out again with the updated library and make this PR `ready for review`.  

Meanwhile, here is a test video that demonstrates the working of new places API. 

https://github.com/user-attachments/assets/fca56d65-7976-4c6f-b54d-a8a80c4c224f

rojiphil committed:
use newer version for the autocomplete library

rojiphil committed:
updated package-lock

rojiphil committed:
lint fix

rojiphil committed:
ts fix

rojiphil committed:
using useOnyx instead of withOnyx - eslint error fix

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rafecolton commented:
Library was published, so placing this in `Ready for review` status

rafecolton reviewed (dismissed):
Tested this out locally and it worked great! @mkhutornyi @rojiphil please complete your cross-platform testing and checklists, and assuming no unexpected errors, I will merge.

Great work!

rojiphil commented:
> please complete your cross-platform testing and checklists, and assuming no unexpected errors, I will merge.

@rafecolton On native iOS session token generation is failing due to unsupported crypto.getRandomValues. Please hold merge until I finish investigation on this.

![Screenshot 2024-10-05 at 1 03 26 AM](https://github.com/user-attachments/assets/58fb8caf-7a47-43b6-9337-0624e393c131)

rafecolton commented:
Is this happening on `main` or is it specific to this PR?

rojiphil commented:
> Is this happening on `main` or is it specific to this PR?

This issue is specific to this PR only. The mains will not be impacted as it still points to 2.5.6 version of library. It is only after we merge this PR that the bumped up 2.5.7 version will kick-in.

rojiphil commented:
> On native iOS session token generation is failing due to unsupported crypto.getRandomValues. 

> That also happens on android.
> Looks like this is the only way to support on native:

I have opened a follow-up PR [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/951) to address this issue. More specific comment [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/951/files#r1789706525) on the fix. I think we can keep this simple by using a random generator for generating the uuid. I have tested and it works for iOS and android native devices.

rojiphil commented:
@rafecolton We have another issue with Android Native platform. This issue and the fix are mentioned [here](https://github.com/FaridSafi/react-native-google-places-autocomplete/pull/951/files#r1789707563). But on adding this, the `desktop` platform fails due to CORS issue. Here is a test video demonstrating the problem. It's strange though that the web platforms do not have any CORS issue. Could adding `content-type` to `Access-Control-Allow-Headers` help solve this issue?

![Screenshot 2024-10-08 at 8 03 15 AM](https://github.com/user-attachments/assets/66a89f88-ba1e-4940-9711-872df4b035cc)

https://github.com/user-attachments/assets/d44b7968-945f-4e27-96dd-b144a83d44db

rafecolton commented:
I did notice the CORS error when testing on dev, so I didn't add the header. It's interesting that it's required for android. I dug into it a bit and it was quite the rabbit hole, but since it seems to be required here, I'll check it out.

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil committed:
Merge branch 'Expensify:main' into new-places-api-support

rojiphil committed:
Merge branch 'main' of https://github.com/Expensify/App into new-places-api-support
